src/Mod/Part/App/Attacher.cpp

```diff
@@ -1185,13 +1186,36 @@ Base::Placement AttachEngine3D::calculateAttachedPlacement(const Base::Placement
         double u, v;
         if (projector.NbPoints()==0)
             throw Base::ValueError("AttachEngine3D::calculateAttachedPlacement: projecting point onto surface failed.");
-        projector.LowerDistanceParameters(u, v);
 
+        projector.LowerDistanceParameters(u, v);
         BRepLProp_SLProps prop(surf,u,v,1, Precision::Confusion());
-        SketchNormal = prop.Normal();
-
         gp_Dir dirX;
-        prop.TangentU(dirX); //if normal is defined, this should be defined too
+
+        if (prop.D1U().Magnitude() > Precision::Confusion() &&
+            prop.D1V().Magnitude() > Precision::Confusion() &&
+            prop.IsNormalDefined()) {
+            SketchNormal = prop.Normal();
+            prop.TangentU(dirX);
+        }
+        // use an alternative method in case of a null normal
+        else {
+            Standard_Boolean done;
+            CSLib_NormalStatus stat;
+            CSLib::Normal(prop.D1U(), prop.D1V(), prop.D2U(), prop.D2V(), prop.DUV(),
+                Precision::Confusion(), done, stat, SketchNormal);
+            if (!done) {
+                throw Base::ValueError("AttachEngine3D::calculateAttachedPlacement: finding normal to surface at projected point failed.");
+            }
+            if (prop.IsTangentUDefined()) {
+                prop.TangentU(dirX);
+            }
+            else {
```

Comment:
Shouldn't you check here if IsTangentVDefined()?

Reply:
The method needs at least one of the tangents (U or V) to be non-zero as a necessary condition to compute a non-zero normal. So if the method successfully finds the normal, if IsTangentUDefined() is false, IsTangentVDefined() is necessarily true.